### ISSUE TITLE
IL2CPP support

### ISSUE THREAD

**Issue by FernandoMonroy** [keijiro/Akvfx#9]
Does Azure support IL2CPP?

**Comment by keijiro** [comment]
Azure? Did you mean Azure Kinect? (Azure is Microsoft's brand name for their cloud computing solutions)

Anyway, this project should support IL2CPP. Please report if you find any issue with IL2CPP.

**Comment by keijiro** [comment]
I'm closing this issue now. Please feel free to reopen for further problems.

**Comment by FernandoMonroy** [comment]
Sure Azure Kinect, sorry for the misleading.

I encounter problems with Intel real sense (using or not the Rsvfx repo) and old Kinect with unity compiling for IL2CPP. The compilations were successful but devices didn´t connect so I see no output of any kind (color or depth); Mono was the only way to run the devices. Is this the case of Azure Kinect? Does Azure Kinect connect properly using the IL2CPP scripting backend?

Don't have an Azure Kinect right now, but I´m considering buying one.

It´ll be super helpful if you could give me some feedback about it.

Thanks for your time.

**Comment by keijiro** [comment]
Thank you for the additional information. I confirmed that the issue reproduces with IL2CPP. You can see the following error message when building with the development build option.

```
NotSupportedException: To marshal a managed method, please add an attribute named 'MonoPInvokeCallback' to the method definition. The method we're attempting to marshal is: Microsoft.Azure.Kinect.Sensor.Logger::OnDebugMessage
  at Microsoft.Azure.Kinect.Sensor.NativeMethods.k4a_set_debug_message_handler (Microsoft.Azure.Kinect.Sensor.NativeMethods+k4a_logging_message_cb_t message_cb, System.IntPtr message_cb_context, Microsoft.Azure.Kinect.Sensor.LogLevel min_level) [0x00000] in <00000000000000000000000000000000>:0 
  at Microsoft.Azure.Kinect.Sensor.Logger.Initialize () [0x00000] in <00000000000000000000000000000000>:0 
  at Microsoft.Azure.Kinect.Sensor.Logger.add_LogMessage (System.Action`1[T] value) [0x00000] in <00000000000000000000000000000000>:0 
  at Microsoft.Azure.Kinect.Sensor.AzureKinectOpenDeviceException.ThrowIfNotSuccess[T] (System.Func`1[TResult] function) [0x00000] in <00000000000000000000000000000000>:0 
  at Microsoft.Azure.Kinect.Sensor.Device.Open (System.Int32 index) [0x00000] in <00000000000000000000000000000000>:0 
  at Akvfx.ThreadedDriver.CaptureThread () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Application+LowMemoryCallback.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at System.Threading.ParameterizedThreadStart.Invoke (System.Object obj) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Application+LowMemoryCallback.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
UnityEngine.Logger:LogException(Exception, Object)
UnityEngine.Debug:LogException(Exception)
System.Xml.Schema.ValidationEventHandler:Invoke(Object, ValidationEventArgs)
```

I found the same issue in discord-rpc:

https://github.com/discordapp/discord-rpc/issues/186

As noted in this page, the issue may be easily solved by adding the `MonoPInvokeCallback` attribute to `Microsoft.Azure.Kinect.Sensor.Logger::OnDebugMessage`. However, it's in the Azure Kinect SDK code, so we have to send a pull request and get reviewed by Microsoft to apply the fix. That's a little bit, uh, troublesome.

At the moment, please understand that this project doesn't support IL2CPP. If you're motivated to solve the issue for some reason, please try the solution above.

**Comment by FernandoMonroy** [comment]
Sorry for the late reply. Thank you very much for your feedback, I'll dig out into the subject

**Comment by mfshiu** [comment]
Yes, switching to Mono scripting backend avoids this issue entirely. Since Mono doesn't require the MonoPInvokeCallback attribute for native callbacks like Logger::OnDebugMessage, you don't need to modify the Azure Kinect SDK or submit a pull request to Microsoft. It's a much simpler workaround.

Certainly! Here's how to configure **Unity to use Mono scripting backend** (instead of IL2CPP):

---

## ✅ How to Set Unity to Use Mono Scripting Backend

1. **Open Unity Editor**

   * Launch your Unity project.

2. **Open Build Settings**

   * Go to `File` > `Build Settings`.

3. **Select Target Platform**

   * Choose your platform (e.g., **PC, Mac & Linux Standalone** > **Windows**).

4. **Click on "Player Settings…"**

   * This opens the **Inspector** window with Player settings.

5. **Find the "Other Settings" Section**

   * In the Inspector window under **Player**, locate the tab for your platform (e.g., Windows, macOS).
   * Expand **Other Settings**.

6. **Change the Scripting Backend**

   * Under **Configuration**, find **Scripting Backend**.
   * Set it to **Mono** (instead of IL2CPP).

7. **Optional: Set API Compatibility Level**

   * Still under **Other Settings**, choose **.NET Standard 2.1** or **.NET 4.x** for **API Compatibility Level**, depending on your plugin requirements.

8. **Close Settings and Build**

   * Close the settings, then go back to `File > Build` to build your project with Mono mode.

---

## 🔍 Notes

* **Mono** is easier for debugging and native plugin interaction (like Azure Kinect), especially when using `DllImport` or external C libraries.
* **IL2CPP** is more optimized and secure, but often has compatibility issues with complex native plugins.
